janitor/functions/bin_numeric.py

```diff
@@ -1,60 +1,66 @@
 import pandas_flavor as pf
 import pandas as pd
 
 from janitor.utils import deprecated_alias
 
 
+ScalarSequence = Sequence[float]
+
+
 @pf.register_dataframe_method
-@deprecated_alias(from_column="from_column_name", to_column="to_column_name")
+@deprecated_alias(
+    from_column="from_column_name",
+    to_column="to_column_name",
+    num_bins="bins",
+)
 def bin_numeric(
     df: pd.DataFrame,
     from_column_name: Hashable,
     to_column_name: Hashable,
-    num_bins: int = 5,
-    labels: Optional[str] = None,
+    bins: Optional[Union[int, ScalarSequence, pd.IntervalIndex]] = 5,
+    **kwargs,
 ) -> pd.DataFrame:
     """
     Generate a new column that labels bins for a specified numeric column.
 
     This method mutates the original DataFrame.
 
-    Makes use of pandas `cut()` function to bin data of one column,
-    generating a new column with the results.
+    A wrapper around the pandas [`cut()`][pd_cut_docs] function to bin data of
+    one column, generating a new column with the results.
+
+    [pd_cut_docs]: https://pandas.pydata.org/docs/reference/api/pandas.cut.html
 
     ```python
-        import pandas as pd
-        import janitor
-        df = (
-            pd.DataFrame(...)
-            .bin_numeric(
-                from_column_name='col1',
-                to_column_name='col1_binned',
-                num_bins=3,
-                labels=['1-2', '3-4', '5-6']
-                )
+    import pandas as pd
+    import janitor
+    df = (
+        pd.DataFrame(...)
+        .bin_numeric(
+            from_column_name='col1',
+            to_column_name='col1_binned',
+            bins=3,
+            labels=['1-2', '3-4', '5-6'],
         )
+    )
     ```
 
     :param df: A pandas DataFrame.
     :param from_column_name: The column whose data you want binned.
     :param to_column_name: The new column to be created with the binned data.
-    :param num_bins: The number of bins to be utilized.
-    :param labels: Optionally rename numeric bin ranges with labels. Number of
-        label names must match number of bins specified.
+    :param bins: The binning strategy to be utilized. Read the `pd.cut`
+        documentation for more details.
+    :param **kwargs: Additional kwargs to pass to `pd.cut`, except `retbins`.
     :return: A pandas DataFrame.
-    :raises ValueError: if number of labels do not match number of bins.
+    :raises ValueError: If `retbins` is passed in as a kwarg.
     """
-    if not labels:
-        df[str(to_column_name)] = pd.cut(
-            df[str(from_column_name)], bins=num_bins
-        )
-    else:
-        if not len(labels) == num_bins:
-            raise ValueError("Number of labels must match number of bins.")
+    if "retbins" in kwargs:
+        raise ValueError("`retbins` is not an acceptable keyword argument.")
 
-        df[str(to_column_name)] = pd.cut(
-            df[str(from_column_name)], bins=num_bins, labels=labels
-        )
+    df[str(to_column_name)] = pd.cut(
```

Comment:
- let's make it such that it does not mutate the original dataframe; `assign` can be helpful here , in which case you use a dictionary within the assign function.
- do we need the `str` function? i'd say it is better to check if the column name is a string with the `check` function in utils, and check if the `from_column_name` exists in the df, with the `check_column` function, also in utils.

Reply:
1. Sure. 2. I suppose not, I just assumed it was cast to str because the original author accepted generic hashables for the input. I'll add in some checks to enforce the arguments as strings and change the docs accordingly.

Reply:
Ahhh I didn’t think of that. If that’s the case maybe do a check for hashable instead? Just not sure how to integrate that into assign

Reply:
My view is to keep it consistent across the library? If other similar functions (e.g. `add_column`) require the "new column name" parameter to be a string (+ if we haven't gotten complaints about not supporting generic hashables), we can stick with string here as well for the time being. :)

Also, according to `add_column` [function docstring](https://pyjanitor-devs.github.io/pyjanitor/api/functions/#janitor.functions.add_columns.add_column) for its `column_name` parameter: 

> Name of the new column. Should be a string, in order for the column name to be compatible with the Feather binary format (this is a useful thing to have).